ktor-io/common/src/io/ktor/utils/io/ByteReadChannelOperations.kt

```diff
@@ -568,53 +563,92 @@ public suspend fun ByteReadChannel.readUntil(
+                "Limit of $limit bytes exceeded " +
+                    "while searching for \"${matchString.toSingleLineString()}\""
+            )
+        }
     }
 
-    fun resetPartialMatch(byte: Byte) {
-        while (matchIndex > 0 && byte != matchString[matchIndex]) {
-            matchIndex = partialMatchTable[matchIndex - 1]
+    // Quick scan for first byte
+    suspend fun advanceToNextMatch() {
+        while (!input.exhausted() || awaitContent()) {
+            val nextMatch = input.indexOf(matchString[0])
+            when {
+                nextMatch == -1L -> {
+                    checkBounds((input as Buffer).size)
+                    rc += input.transferTo(writeChannel.writeBuffer)
+                }
+                else -> {
+                    checkBounds(nextMatch)
+                    rc += input.readAtMostTo(writeChannel.writeBuffer as Buffer, nextMatch)
+                    return
+                }
+            }
         }
     }
 
-    while (!isClosedForRead) {
-        val byte = readByte()
+    // Slow sequential check for potential match
+    suspend fun readCompleteMatch(): Boolean {
+        while (!input.exhausted() || awaitContent()) {
+            val byte = input.readByte()
 
-        if (matchIndex > 0 && byte != matchString[matchIndex]) {
-            appendPartialMatch()
-            resetPartialMatch(byte)
-        }
+            if (matchIndex > 0 && byte != matchString[matchIndex]) {
+                // update match index from our table
+                val oldMatchIndex = matchIndex
+                while (matchIndex > 0 && byte != matchString[matchIndex]) {
+                    matchIndex = partialMatchTable[matchIndex - 1]
+                }
+                // write the discarded partial match
+                val retained = (oldMatchIndex - matchIndex).toLong()
+                checkBounds(retained)
+                rc += partialMatchBuffer.readAtMostTo(
+                    writeChannel.writeBuffer as Buffer,
+                    retained
+                )
+                // no longer matching, scan for next match
+                if (matchIndex == 0 && byte != matchString[matchIndex]) {
+                    writeChannel.writeByte(byte)
+                    rc++
+                    return false
+                }
+            }
 
-        if (byte == matchString[matchIndex]) {
-            matchBuffer[matchIndex] = byte
+            // return if complete match, else add to partial buffer
             if (++matchIndex == matchString.size) {
-                return rc
+                return true
             }
-        } else {
-            writeChannel.writeByte(byte)
-            rc++
+            partialMatchBuffer.writeByte(byte)
         }
+        return false
+    }
 
-        if (rc > limit) {
-            throw IOException("Limit of $limit bytes exceeded while scanning for \"${matchString.decodeToString()}\"")
+    while (!input.exhausted() || awaitContent()) {
```

Comment:
Is it intended that the code in the `advanceToNextMatch` and `readCompleteMatch` functions have the same condition in the while loops?